qldpc/stim/memory.py

```diff
@@ -0,0 +1,282 @@
+    """
+    circuit.append(sm_circuit)
+    for meas_round in sm_measurements:
+        _update_meas_rec(meas_rec, meas_round)
+    if basis is Pauli.X:
+        for i, check_id in enumerate(qubit_ids.x_check):
+            circuit.append("DETECTOR", [stim.target_rec(meas_rec[-1][check_id])], (1, i, 0))
+    elif basis is Pauli.Z:
+        for i, check_id in enumerate(qubit_ids.z_check):
+            circuit.append("DETECTOR", [stim.target_rec(meas_rec[-1][check_id])], (2, i, 0))
+
+    """
+    Repeated syndrome rounds
+    """
+    repeat_circuit = stim.Circuit()
+    repeat_circuit.append(sm_circuit)
+    for meas_round in sm_measurements:
+        _update_meas_rec(meas_rec, meas_round)
+    for i, check_id in enumerate(qubit_ids.x_check):
+        repeat_circuit.append(
+            "DETECTOR",
+            [
+                stim.target_rec(meas_rec[-1][check_id]),
+                stim.target_rec(meas_rec[-2][check_id]),
+            ],
+            (1, i, 1),
+        )
+    for i, check_id in enumerate(qubit_ids.z_check):
+        repeat_circuit.append(
+            "DETECTOR",
+            [
+                stim.target_rec(meas_rec[-1][check_id]),
+                stim.target_rec(meas_rec[-2][check_id]),
+            ],
+            (2, i, 1),
+        )
+    repeat_circuit.append("SHIFT_COORDS", [], (0, 0, 1))
+    circuit.append(stim.CircuitRepeatBlock(num_rounds - 1, repeat_circuit))
+
+    """
+    Measure out data qubits
+    """
+    if basis is Pauli.X:
+        circuit.append("MX", qubit_ids.data)
+    elif basis is Pauli.Z:
+        circuit.append("M", qubit_ids.data)
+
+    _update_meas_rec(meas_rec, qubit_ids.data)
+
+    """
+    Reconstruct a final round of checks based on data qubit measurements
+    """
+    if basis is Pauli.X:
+        for i, check_id in enumerate(qubit_ids.x_check):
+            data_support = np.where(code.code_x.matrix[i])[0]
+            circuit.append(
+                "DETECTOR",
+                [stim.target_rec(meas_rec[-1][qubit_ids.data[q]]) for q in data_support]
+                + [stim.target_rec(meas_rec[-2][check_id])],
+                (1, i, num_rounds),
+            )
+    elif basis is Pauli.Z:
+        for i, check_id in enumerate(qubit_ids.z_check):
+            data_support = np.where(code.code_z.matrix[i])[0]
+            circuit.append(
+                "DETECTOR",
+                [stim.target_rec(meas_rec[-1][qubit_ids.data[q]]) for q in data_support]
+                + [stim.target_rec(meas_rec[-2][check_id])],
+                (2, i, num_rounds),
+            )
+
+    """
+    Define observables for memory experiment
+    """
+    observables = code.get_logical_ops(basis)
+    for k, obs in enumerate(observables):
+        data_support = np.where(obs)[0]
+        circuit.append(
+            "OBSERVABLE_INCLUDE",
+            [stim.target_rec(meas_rec[-1][qubit_ids.data[q]]) for q in data_support],
+            k,
+        )
```

Comment:
Given that data qubits were already measured either in the X or Z basis, I think only the corresponding X or Z observables are here valid.